pinot-server/src/main/java/org/apache/pinot/server/starter/helix/SendStatsPredicate.java

```diff
@@ -0,0 +1,160 @@
+ * under the License.
+ */
+package org.apache.pinot.server.starter.helix;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import org.apache.helix.NotificationContext;
+import org.apache.helix.api.listeners.InstanceConfigChangeListener;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.pinot.common.version.PinotVersion;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class SendStatsPredicate implements InstanceConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SendStatsPredicate.class);
+
+  public abstract boolean getSendStats();
+
+  public static SendStatsPredicate create(PinotConfiguration configuration) {
+    String modeStr = configuration.getProperty(
+        CommonConstants.MultiStageQueryRunner.KEY_OF_SEND_STATS_MODE,
+        CommonConstants.MultiStageQueryRunner.DEFAULT_SEND_STATS_MODE).toUpperCase(Locale.ENGLISH);
+    for (Mode mode : Mode.values()) {
+      if (mode.name().equals(modeStr)) {
+        return mode.create();
+      }
+    }
+
+    throw new IllegalStateException("Invalid value " + modeStr + " for "
+        + CommonConstants.MultiStageQueryRunner.KEY_OF_SEND_STATS_MODE);
+  }
+
+  public enum Mode {
+    SAFE {
+      @Override
+      public SendStatsPredicate create() {
+        return new Safe();
+      }
+    },
+    ALWAYS {
+      @Override
+      public SendStatsPredicate create() {
+        return new SendStatsPredicate() {
+          @Override
+          public boolean getSendStats() {
+            return false;
+          }
+
+          @Override
+          public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+            // Nothing to do
+          }
+        };
+      }
+    },
+    NEVER {
+      @Override
+      public SendStatsPredicate create() {
+        return new SendStatsPredicate() {
+          @Override
+          public boolean getSendStats() {
+            return false;
+          }
+
+          @Override
+          public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+            // Nothing to do
+          }
+        };
+      }
+    };
+
+    public abstract SendStatsPredicate create();
+  }
+
+  private static class Safe extends SendStatsPredicate {
+    private final AtomicBoolean _sendStats = new AtomicBoolean(true);
+
+    @Override
+    public boolean getSendStats() {
+      return _sendStats.get();
+    }
+
+    @Override
+    public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+      boolean sendStats = true;
+      // TODO: Right now we read all configs, while only broker and servers are actually important
```

Comment:
You should be able to use `InstanceTypeUtils`

Reply:
Nice! I'm changing it